### ISSUE TITLE
How to convert the motion files obtained by TRAM to the smpl motion files for the ProtoMotions project

### ISSUE THREAD

**Issue by yangyixiang-cc** [LeCAR-Lab/ASAP#57]
![Image](https://github.com/user-attachments/assets/6ae7e3f6-1e81-4459-830b-f17f834d9ec1)
Hello dear developers, congratulations on developing an amazing project, I encountered difficulties in reproducing, mainly in the B->C phase as shown below, how to convert the motion files obtained by TRAM to the smpl motion files for the ProtoMotions project. I'd appreciate it if you could make this part of the conversion code public.

**Comment by Ce1estenoob** [comment]
Have you solved this problem? I met this issue while reproducing, too.

**Comment by yangyixiang-cc** [comment]
@Ce1estenoob Here's my implementation to convert the hps file .npy of the tram output to amass standard format .npz, here's the code: 
`
"""
将TRAM模型直接输出的NPY数据转换为标准SMPL格式 (.npz) 的脚本。

输入: TRAM HPS结果文件 (*.npy)
    - pred_rotmat: (T, 24, 3, 3) - 旋转矩阵格式的姿态
    - pred_shape: (T, 10) - 逐帧的身体形状参数
    - pred_trans: (T, 3) - 全局位移
    - frame: (T,) - 帧索引

输出: 标准SMPL格式文件 (*.npz)
    - poses: (T, 72) - 轴角表示的身体姿态 (SMPL标准)
    - betas: (10,) - 平均的身体形状参数 (SMPL标准)
    - trans: (T, 3) - 全局位移
    - gender: 'male', 'female', 或 'neutral'
    - mocap_framerate: 帧率
    - frame_indices: 原始帧索引
"""

import sys
import os
sys.path.insert(0, os.path.dirname(__file__) + '/..')

import numpy as np
import torch
import argparse
from glob import glob
from tqdm import tqdm

try:
    from lib.utils.rotation_conversions import matrix_to_axis_angle
except ImportError:
    print("错误: 无法导入 matrix_to_axis_angle。")
    sys.exit(1)


def convert_single_file(npy_file, output_file=None, gender='neutral', framerate=30.0, isaac_compat=False):
    """
    转换单个TRAM npy文件到SMPL npz格式。

    Args:
        npy_file (str): 输入的TRAM HPS .npy文件路径。
        output_file (str, optional): 输出的.npz文件路径。如果为None，则在原位置生成。
        gender (str, optional): 模型性别。
        framerate (float, optional): 动作帧率。
        isaac_compat (bool, optional): 是否为Isaac Sim进行坐标系转换 (Y-up -> Z-up)。
    """
    print(f"--> 正在处理文件: {os.path.basename(npy_file)}")
    if isaac_compat:
        print("    - 已启用Isaac Sim兼容模式 (Z-up 坐标系转换)。")

    if output_file is None:
        output_file = npy_file.replace('.npy', '_smpl.npz')

    # 1. 加载TRAM HPS结果
    try:
        tram_data = np.load(npy_file, allow_pickle=True).item()
    except Exception as e:
        print(f"    [错误] 无法加载文件 {npy_file}: {e}")
        return

    # 2. 提取数据
    pred_rotmat = tram_data.get('pred_rotmat')  # (T, 24, 3, 3)
    pred_shape = tram_data.get('pred_shape')    # (T, 10)
    pred_trans = tram_data.get('pred_trans')    # (T, 3)
    frame_indices = tram_data.get('frame')      # (T,)

    if pred_rotmat is None or pred_shape is None or pred_trans is None:
        print(f"    [错误] 文件 {npy_file} 缺少 'pred_rotmat', 'pred_shape', 或 'pred_trans' 键。")
        return

    T = len(pred_rotmat)
    print(f"    - 总帧数: {T}")

    # 3. 转换数据到SMPL标准格式
    # 3.1. 姿态：将旋转矩阵 (T, 24, 3, 3) 转换为轴角 (T, 72)
    if isinstance(pred_rotmat, np.ndarray):
        pred_rotmat = torch.from_numpy(pred_rotmat).float()
    
    # === Isaac Sim 兼容性转换 ===
    if isaac_compat:
        # 定义从 Y-up 到 Z-up 的旋转矩阵 (绕X轴旋转-90度)
        angle = -torch.pi / 2.0
        cos_a = np.cos(angle)
        sin_a = np.sin(angle)
        r_x = torch.tensor([
            [1, 0, 0],
            [0, cos_a, -sin_a],
            [0, sin_a,  cos_a]
        ], dtype=pred_rotmat.dtype, device=pred_rotmat.device)

        # 1. 转换根关节旋转
        root_rot = pred_rotmat[:, 0, :, :]
        new_root_rot = torch.matmul(r_x.unsqueeze(0), root_rot)
        pred_rotmat[:, 0, :, :] = new_root_rot
        
        # 2. 转换全局位移
        if pred_trans is not None:
            if isinstance(pred_trans, np.ndarray):
                pred_trans = torch.from_numpy(pred_trans).float()
            
            # 确保 pred_trans 是 (T, 3) 的形状
            squeezed_trans = pred_trans.squeeze()
            if squeezed_trans.dim() == 2 and squeezed_trans.shape[1] == 3:
                x, y, z = squeezed_trans[:, 0], squeezed_trans[:, 1], squeezed_trans[:, 2]
                pred_trans = torch.stack([x, -z, y], dim=1)
            else:
                 print(f"    [警告] 'pred_trans' 的形状不符合预期 (实际为 {pred_trans.shape})，跳过全局位移的坐标系转换。")


    rotmat_flat = pred_rotmat.reshape(-1, 3, 3)
    axis_angle_flat = matrix_to_axis_angle(rotmat_flat)
    smpl_poses = axis_angle_flat.reshape(T, -1).cpu().numpy()

    # 3.2. 体型：计算逐帧体型参数的平均值，得到固定体型
    if isinstance(pred_shape, torch.Tensor):
        pred_shape = pred_shape.cpu().numpy()
    smpl_betas = np.mean(pred_shape, axis=0)

    # 3.3. 位移：直接使用
    if isinstance(pred_trans, torch.Tensor):
        pred_trans = pred_trans.cpu().numpy()
    smpl_trans = pred_trans.squeeze()

    # 4. 组装SMPL格式字典
    smpl_data = {
        'poses': smpl_poses,           # (T, 72) - 轴角
        'betas': smpl_betas,           # (10,) - 形状
        'trans': smpl_trans,           # (T, 3) - 位移
        'gender': str(gender),
        'mocap_framerate': float(framerate),
    }
    if frame_indices is not None:
        if isinstance(frame_indices, torch.Tensor):
            frame_indices = frame_indices.cpu().numpy()
        smpl_data['frame_indices'] = frame_indices

    # 5. 保存为 .npz 文件
    try:
        np.savez(output_file, **smpl_data)
        print(f"    - 转换成功! 已保存至: {output_file}")
    except Exception as e:
        print(f"    [错误] 保存文件 {output_file} 失败: {e}")


def convert_folder(input_dir, output_dir=None, gender='neutral', framerate=30.0, isaac_compat=False):
    """
    转换一个文件夹中所有的TRAM .npy文件。

    Args:
        input_dir (str): 包含.npy文件的输入文件夹。
        output_dir (str, optional): 保存.npz文件的输出文件夹。如果为None，则在原文件夹中创建。
        gender (str, optional): 模型性别。
        framerate (float, optional): 动作帧率。
        isaac_compat (bool, optional): 是否为Isaac Sim进行坐标系转换。
    """
    # 优先在 'hps' 子目录中查找HPS文件
    hps_dir = os.path.join(input_dir, 'hps')
    search_dir = input_dir
    
    if os.path.isdir(hps_dir):
        print(f"检测到 'hps' 子目录，将在 {hps_dir} 中搜索HPS文件。")
        search_dir = hps_dir
    else:
        print(f"未找到 'hps' 子目录，将在主结果目录 {input_dir} 中搜索。")

    npy_files = sorted(glob(os.path.join(search_dir, '*.npy')))
    if not npy_files:
        print(f"错误: 在目录 {search_dir} 中未找到任何 .npy 文件。")
        print("请确认 'estimate_humans.py' 已成功运行并在此目录中生成了HPS结果文件。")
        sys.exit(1)

    print(f"\n在 {search_dir} 中找到 {len(npy_files)} 个 .npy 文件，开始批量转换...")

    if output_dir:
        os.makedirs(output_dir, exist_ok=True)
        print(f"输出目录: {output_dir}")
    
    successful_conversions = 0
    for npy_file in tqdm(npy_files, desc="转换进度"):
        if output_dir:
            base_name = os.path.basename(npy_file).replace('.npy', '_smpl.npz')
            output_file = os.path.join(output_dir, base_name)
        else:
            # 如果输出目录未指定，则在输入文件旁边创建
            output_file = npy_file.replace('.npy', '_smpl.npz')
        
        try:
            convert_single_file(npy_file, output_file, gender, framerate, isaac_compat)
            successful_conversions += 1
        except Exception as e:
            print(f"处理文件 {npy_file} 时发生意外错误: {e}")

    if successful_conversions == 0 and len(npy_files) > 0:
        print("\n错误: 所有文件都转换失败！")
        sys.exit(1)
    elif successful_conversions < len(npy_files):
        print(f"\n警告: {len(npy_files) - successful_conversions} 个文件转换失败。")


if __name__ == "__main__":
    parser = argparse.ArgumentParser(
        description='将TRAM模型输出的NPY转换为标准SMPL格式的NPZ文件。',
        formatter_class=argparse.RawTextHelpFormatter
    )
    parser.add_argument('input_path', type=str,
                        help='输入的.npy文件路径或包含.npy文件的文件夹路径。')
    parser.add_argument('--output_path', type=str, default=None,
                        help='输出的.npz文件路径或输出文件夹路径。\n'
                             '如果输入是文件，此项为输出文件名；\n'
                             '如果输入是文件夹，此项为输出文件夹名。\n'
                             '如果留空，则在输入文件相同位置生成。')
    parser.add_argument('--gender', type=str, default='neutral',
                        choices=['male', 'female', 'neutral'],
                        help='设置SMPL模型的性别 (默认: neutral)。')
    parser.add_argument('--framerate', type=float, default=30.0,
                        help='设置动作的帧率 (默认: 30.0)。')
    parser.add_argument('--isaac_compat', action='store_true',
                        help='为Isaac Sim进行坐标系转换 (Y-up -> Z-up)。')

    args = parser.parse_args()

    if os.path.isfile(args.input_path):
        convert_single_file(args.input_path, args.output_path, args.gender, args.framerate, args.isaac_compat)
    elif os.path.isdir(args.input_path):
        convert_folder(args.input_path, args.output_path, args.gender, args.framerate, args.isaac_compat)
    else:
        print(f"错误: 输入路径不存在 -> {args.input_path}")
        sys.exit(1)

    print("\n转换完成。") ` 
这是运行命令：python scripts/convert_tram_to_smpl.py "$HPS_RESULT_DIR" --isaac_compat
